### DISCUSSION TITLE
What listeners should be considered when filtering hostnames in routes?

### DISCUSSION THREAD

**Discussion by randmonkey** [kubernetes-sigs/gateway-api#1563] [General]
Currently, `HTTPRoute`, `TLSRoute`, `GRPCRoute` includes hostnames in their specifications. As described in [API specifiations](https://gateway-api.sigs.k8s.io/references/spec/), their hostnames should be filtered to include only the **intersection** of hostnames  in its attached listeners. So what listeners should be taken into consideration if the route has not specified a listener attached to it? I have came out of the following options:
- The implementation MUST attach exactly one listener to route if there are at least one listener can be attach to it, so that we only consider the attached listener.
- Only listeners with compatible protocol with the route are considered:
  -  `HTTP` and `HTTPS` (`TLS`?) listeners for `HTTPRoute`
  - `TLS` listeners for `TLSRoute`
  - `HTTPS` (`HTTP`? `TLS`?) listeners for `GRPCRoute` .
- ALL listeners that are possible to set a hostname are considered. This includes all listeners with `HTTP`, `HTTPS`, or`TLS` protocol.

**Comment by youngnick** [comment]
For the route to be useful, it has to attach to a Gateway; if a specific listener is not selected, the Route should attach to *all* compatible listeners in that Gateway.

Then, implementations should use "no requested hostname == match *any* hostname" for both sides of the Gateway and Route match.

This could probably use a small spec update to clarify though.

Also, we probably haven't done enough to give specific requirements about what Protocol matches what Route types though.

**Reply by randmonkey** [reply]
Thank you. I also think option 2: attach to **ALL compatible** listeners is the most reasonable answer. But It was not specified that what protocols of listeners could be compatible with which type of route.

**Reply by youngnick** [reply]
I think that for protocol support it's kind of left as an exercise for implementations for now, but I don't think that's enough to ensure compatibility for the future. I'll make a note.
